druid-shell/src/platform/gtk/window.rs

```diff
@@ -832,24 +834,62 @@ fn get_mouse_click_count(event_type: gdk::EventType) -> u8 {
     }
 }
 
-fn get_modifiers(modifiers: gdk::ModifierType) -> keyboard::KeyModifiers {
-    keyboard::KeyModifiers {
-        shift: modifiers.contains(ModifierType::SHIFT_MASK),
-        alt: modifiers.contains(ModifierType::MOD1_MASK),
-        ctrl: modifiers.contains(ModifierType::CONTROL_MASK),
-        meta: modifiers.contains(ModifierType::META_MASK),
+const MODIFIER_MAP: &[(gdk::ModifierType, Modifiers)] = &[
+    (ModifierType::SHIFT_MASK, Modifiers::SHIFT),
+    (ModifierType::MOD1_MASK, Modifiers::ALT),
+    (ModifierType::CONTROL_MASK, Modifiers::CONTROL),
+    (ModifierType::META_MASK, Modifiers::META),
+];
+
+fn get_modifiers(modifiers: gdk::ModifierType) -> Modifiers {
+    let mut result = Modifiers::empty();
+    for &(gdk_mod, modifier) in MODIFIER_MAP {
+        if modifiers.contains(gdk_mod) {
+            result |= modifier;
+        }
     }
+    result
 }
 
-fn make_key_event(key: &EventKey, repeat: bool) -> keyboard::KeyEvent {
+fn make_key_event(key: &EventKey, repeat: bool, state: KeyState) -> KeyboardEvent {
     let keyval = key.get_keyval();
     let hardware_keycode = key.get_hardware_keycode();
 
     let keycode = hardware_keycode_to_keyval(hardware_keycode).unwrap_or(keyval);
 
     let text = gdk::keyval_to_unicode(keyval);
-
-    keyboard::KeyEvent::new(keycode, repeat, get_modifiers(key.get_state()), text, text)
+    /*
+    println!(
+        "keyval = 0x{:x} ({}), hw = 0x{:x}, text = {:?}",
+        keyval, keyval, hardware_keycode, text
+    );
+    */
+    let modifiers = get_modifiers(key.get_state());
+    let key = keycodes::raw_key_to_key(keyval).unwrap_or_else(|| {
+        if let Some(c) = text {
+            if c >= ' ' && c != '\x7f' {
+                Key::Character(c.to_string())
+            } else {
+                Key::Unidentified
+            }
+        } else {
+            Key::Unidentified
+        }
+    });
+    let code = keycodes::hardware_keycode_to_code(hardware_keycode);
+    let location = keycodes::raw_key_to_location(keycode);
+    let is_composing = false;
+    // TODO: set this to "up" for key up events
```

Comment:
this is future work?

Reply:
Nah, I did it and forgot to remove the TODO.